File: ffi/ts/src/subAddrId.ts
import {
  castToSubAddrId,
  deserializeSubAddrId,
  genSubAddrId,
  serializeSubAddrId,
} from './blsct'

import { ManagedObj } from './managedObj'

/** Represents a sub-address ID.
 *
 * Examples:
 * ```ts
 * const { SubAddrId } = require('navio-blsct')
 * const x = SubAddrId.generate(123, 456)
 * x
 * const ser = x.serialize()
 * const deser = SubAddrId.deserialize(ser)
 * ser === deser.serialize() // true
 * ```
 */
export class SubAddrId extends ManagedObj {
  constructor(obj: any) {
    super(obj)
  }

  /** Generates a new `SubAddrId` instance.
   *
   * @param account - The account for the sub-address.
   * @param address - The address for the sub-address.
   * @return A new `SubAddrId` instance with the specified account and address.
   */
  static generate(
    account: number,
    address: number,
  ): SubAddrId {
    const obj = genSubAddrId(account, address)
    return new SubAddrId(obj)
  }

  override value(): any {
    return castToSubAddrId(this.obj)
  }

  override serialize(): string {
    return serializeSubAddrId(this.value())
  }

  /** Deserializes a `SubAddrId` from its hexadecimal representation.
   *
   * @param hex - The hexadecimal string to deserialize.
   * @returns A new `SubAddrId` instance.
   */
  static deserialize(
    this: new (obj: any) => SubAddrId,
    hex: string
  ): SubAddrId {
    return SubAddrId._deserialize(hex, deserializeSubAddrId)
  }
}

File: ffi/rust/src/sub_addr_id.rs
use crate::{
  blsct_obj::BlsctObj,
  blsct_serde::BlsctSerde, 
  ffi::{
    BlsctSubAddrId,
    BlsctRetVal,
    deserialize_sub_addr_id,
    gen_sub_addr_id,
    get_sub_addr_id_account,
    get_sub_addr_id_address,
    serialize_sub_addr_id,
  },
  macros::{
    impl_clone,
    impl_display,
    impl_from_retval,
    impl_value,
  },
};
use serde::{Deserialize, Serialize};
use std::ffi::c_char;

#[derive(Debug, Deserialize, Serialize, Eq)]
pub struct SubAddrId {
  obj: BlsctObj<SubAddrId, BlsctSubAddrId>,
}

impl_from_retval!(SubAddrId);
impl_display!(SubAddrId);
impl_clone!(SubAddrId);

impl SubAddrId {
  pub fn new(
    account: i64,
    address: u64,
  ) -> Self {
    let blsct_key_id = unsafe { gen_sub_addr_id(account, address) };
    BlsctObj::from_c_obj(blsct_key_id).into()
  }

  pub fn account(self) -> i64 {
    unsafe { get_sub_addr_id_account(self.value()) }
  }

  pub fn address(self) -> u64 {
    unsafe { get_sub_addr_id_address(self.value()) }
  }

  impl_value!(SubAddrId, BlsctSubAddrId);
}

impl BlsctSerde for SubAddrId {
  unsafe fn serialize(ptr: *const u8, _: usize) -> *const i8 {
    serialize_sub_addr_id(ptr as *const BlsctSubAddrId)
  }

  unsafe fn deserialize(hex: *const c_char) -> *mut BlsctRetVal {
    deserialize_sub_addr_id(hex)
  }
}

impl PartialEq for SubAddrId {
  fn eq(&self, other: &Self) -> bool {
    self.obj == other.obj 
  }
}

impl From<BlsctObj<SubAddrId, BlsctSubAddrId>> for SubAddrId {
  fn from(obj: BlsctObj<SubAddrId, BlsctSubAddrId>) -> SubAddrId {
    SubAddrId { obj }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::initializer::init;

  #[test]
  fn test_account() {
    init();
    let a = SubAddrId::new(123, 456);
    assert_eq!(a.account(), 123);
  }

  #[test]
  fn test_address() {
    init();
    let a = SubAddrId::new(123, 456);
    assert_eq!(a.address(), 456);
  }

  #[test]
  fn test_eq() {
    init();
    let a = SubAddrId::new(123, 456);
    let b = SubAddrId::new(234, 567);

    assert!(a == a);
    assert!(a != b);
    assert!(b != a);
    assert!(b == b);
  }

  #[test]
  fn test_deser() {
    init();
    let a = SubAddrId::new(123, 456);
    let hex = bincode::serialize(&a).unwrap();
    let b = bincode::deserialize::<SubAddrId>(&hex).unwrap();
    assert_eq!(a, b);
  }
}

